src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -517,41 +534,35 @@ internal virtual void AddHeaders(HttpHeaders sourceHeaders)
             Debug.Assert(sourceHeaders != null);
             Debug.Assert(GetType() == sourceHeaders.GetType(), "Can only copy headers from an instance of the same type.");
 
-            Dictionary<HeaderDescriptor, object>? sourceHeadersStore = sourceHeaders._headerStore;
-            if (sourceHeadersStore is null || sourceHeadersStore.Count == 0)
-            {
-                return;
-            }
-
-            _headerStore ??= new Dictionary<HeaderDescriptor, object>();
-
-            foreach (KeyValuePair<HeaderDescriptor, object> header in sourceHeadersStore)
+            foreach (HeaderEntry entry in sourceHeaders.GetEntries())
```

Comment:
If sourceHeaders is backed by a dictionary, this is first going to copy the data into an array, only to then enumerate that array and throw it away.  Can we do better?

Separately but related, how common is it for AddHeaders to be called on an empty collection?  I'm wondering if we can optimize the transfer from sourceHeaders to this collection in some common cases by cloning the whole data structure rather than by adding each header individually.

Reply:
> If sourceHeaders is backed by a dictionary, this is first going to copy the data into an array, only to then enumerate that array and throw it away. Can we do better?

We could, but it would mean either duplicating some logic in callers / incurring the overhead for a custom enumerator in the common case. Related: https://github.com/dotnet/runtime/pull/62981#discussion_r785305903

> how common is it for AddHeaders to be called on an empty collection?

`AddHeaders` in mainly used in two places:
1. Copying `defaultRequestHeaders` to the request in `HttpClient`. I'm not sure how common it is to not specify any headers on the request, but this sounds like a nice optimization we should just do here if the target is empty.
2. Creating a new `HttpContent` in `DecompressionHandler` and copying all the headers. In that case the destination always starts out as empty. I opened #63632 for this case since we could avoid copying altogether.

If it turns out it's really common to have a lot of `defaultRequestHeaders` and few headers on the request, we could optimize for that too in the future.